### PR TITLE
Some automated conversions in saveat

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
This makes it nicer to use Float32 from R and Python, since you can just `saveat=0.01` and it attempts to convert the type to Float32 if the JIT was to Float32

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
@utkarsh530 can you check what's going on with Metal?

**Comment by utkarsh530** [comment]
It should be ready now after the tests pass.

File: test/Project.toml [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
 BenchmarkTools = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
-CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
 DiffEqDevTools = "f3b72e0c-5b89-59e1-b016-84e28bfd966d"
 Distributed = "8ba89e20-285c-5b6f-9357-94700520ee1b"
 Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 GPUArraysCore = "46192b85-c4d5-4398-a991-12ede77f4527"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Metal = "dde4c033-4e86-420c-a63e-0dd931031962"
```
Comment (ChrisRackauckas) [inline test/Project.toml:10]: is this intentional?

File: test/gpu_kernel_de/conversions.jl [hunk 2]
```diff
@@ -17,6 +17,31 @@ p = [10.0f0, 28.0f0, 8 / 3.0f0]
 prob = ODEProblem{false}(lorenz, u0, tspan, p)
 prob_func = (prob, i, repeat) -> remake(prob, p = (@SVector rand(Float32, 3)) .* p)
 monteprob = EnsembleProblem(prob, prob_func = prob_func, safetycopy = false)
-sol = solve(monteprob, GPUTsit5(), EnsembleGPUKernel(backend),
+
+## Don't test the problems in which GPUs don't support FP64 completely yet
+## Creating StepRangeLen causes some param types to be FP64 inferred by `float` function
```
Comment (ChrisRackauckas) [inline test/gpu_kernel_de/conversions.jl:22]: the conversion should change things to Float32?

File: src/ensemblegpukernel/kernels.jl [hunk 3]
```diff
@@ -0,0 +1,114 @@
+
+# saveat is just a bool here:
```
Comment (ChrisRackauckas) [inline src/ensemblegpukernel/kernels.jl:2]: That's not allowed?

**Comment by ChrisRackauckas** [comment]
Merging so I can use it. But I'm still not sure about the bool dispatch.

**Pull request merged**